src/component/handlers/composition/DOMObserver.js

```diff
@@ -0,0 +1,128 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @format
+ * @flow strict-local
+ * @emails oncall+draft_js
+ */
+
+'use strict';
+
+const UserAgent = require('UserAgent');
+const Immutable = require('immutable');
+
+const findAncestorOffsetKey = require('findAncestorOffsetKey');
+const invariant = require('invariant');
+const nullthrows = require('nullthrows');
+
+const {Map} = Immutable;
+
+type MutationRecordT =
+  | MutationRecord
+  | {|type: 'characterData', target: Node, removedNodes?: void|};
+
+// Heavily based on Prosemirror's DOMObserver https://github.com/ProseMirror/prosemirror-view/blob/master/src/domobserver.js
+
+const DOM_OBSERVER_OPTIONS = {
+  subtree: true,
+  characterData: true,
+  childList: true,
+  characterDataOldValue: false,
+  attributes: false,
+};
+// IE11 has very broken mutation observers, so we also listen to DOMCharacterDataModified
+const USE_CHAR_DATA = UserAgent.isBrowser('IE <= 11');
+
+class DOMObserver {
+  observer: MutationObserver;
+  container: HTMLElement;
+  mutations: Map<string, string>;
+  onCharData: ({target: EventTarget, type: string}) => void;
+
+  constructor(container: HTMLElement) {
+    this.container = container;
+    this.mutations = Map();
+    if (window.MutationObserver && !USE_CHAR_DATA) {
+      this.observer = new window.MutationObserver(mutations =>
+        this.registerMutations(mutations),
+      );
+    } else {
+      this.onCharData = e => {
+        invariant(e.target instanceof Node);
```

Comment:
should there be a second argument for `invariant` here? perhaps something like "expected target to be an instance of Node"?

Reply:
done